### PR TITLE
Metrics fetcher for jenkins and code coverage

### PR DESCRIPTION
**Pull request by yutongz**

### CODE REVIEW

File: toolbox/metrics/BUILD [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        "@com_github_prometheus_client_golang//prometheus/promhttp:go_default_library",
+        "@com_google_cloud_go//storage:go_default_library",
+        "@org_golang_x_net//context:go_default_library",
+    ],
+)
+
+go_binary(
+    name = "fetcher",
+    gc_linkopts = [
+        "-linkmode",
```
Comment (sebastienvas) [inline toolbox/metrics/BUILD:19]: Do you know what are those for ?

File: toolbox/metrics/main.go [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+		"test-infra":  {"presubmit"},
+	}
+
+	codeCovRegister = map[string]bool{
+		"pilot": true,
+		"mixer": true,
+		"auth":  true,
+	}
+
+	codeCovTrackJob = "presubmit"
```
Comment (sebastienvas) [inline toolbox/metrics/main.go]: Dont you want to use postsubmit here ? Or actually fastforward job ? presubmit might not get in.

File: toolbox/metrics/main.go [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+		client:      &http.Client{},
+	}
+}
+
+func newRepo(name string) *repo {
+	r := &repo{
+		name: name,
+		jobs: make([]*job, 0),
+	}
+	for n := range jobRegister[name] {
```
Comment (sebastienvas) [inline toolbox/metrics/main.go]: what if name is not in the map, would 'nt you get a nil ?

File: toolbox/metrics/main.go [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+
+	codeCoverage = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "package_code_coverage",
+			Help: "Package code coverage.",
+		},
+		[]string{"package", "repo"},
+	)
+
+	jobRegister = map[string][]string{
```
Comment (sebastienvas) [inline toolbox/metrics/main.go]: Can we create this list dynamically ? Does the API allow job folder listing ?

File: toolbox/metrics/main.go [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+		"pilot":       {"presubmit", "postsubmit", "e2e-smoketest", "e2e-suite"},
+		"mixer":       {"presubmit", "postsubmit", "e2e-smoketest", "e2e-suite"},
+		"proxy":       {"presubmit", "postsubmit"},
+		"auth":        {"presubmit", "postsubmit"},
+		"mixerclient": {"presubmit"},
+		"istio":       {"presubmit"},
+		"test-infra":  {"presubmit"},
+	}
+
+	codeCovRegister = map[string]bool{
```
Comment (sebastienvas) [inline toolbox/metrics/main.go]: Same comment here, Can we attempt to get code coverage on all the repos and just print 0 on others ?

File: toolbox/metrics/main.go [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+	return nil
+}
+
+func (j *job) getLatestCompletedBuild() (int, error) {
+	o, err := j.getJenkinsObject("lastCompletedBuild")
+	if err != nil {
+		return 0, fmt.Errorf("failed to get last build info of build, %v", err)
+	}
+	if len(o) == 0 {
+		return 0, errors.New("get empty result from jenkins")
```
Comment (sebastienvas) [inline toolbox/metrics/main.go]: got empty results ...

File: toolbox/metrics/main.go (hunk 3 above)
Comment (yutongz) [reply toolbox/metrics/main.go]: yes, we get a nil if the key doesn't exist, and when n == nil, it won't go into the loop

**Review by sebastienvas — CHANGES_REQUESTED** [review]
PTAL

File: toolbox/metrics/main.go (hunk 2 above)
Comment (yutongz) [reply toolbox/metrics/main.go]: let me use postsubmit for now.

File: toolbox/metrics/BUILD (hunk 1 above)
Comment (yutongz) [reply toolbox/metrics/BUILD:19]: static build is required for building docker image

**Comment by yutongz** [comment]
PTAL

File: toolbox/metrics/main.go [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+		prometheus.GaugeOpts{
+			Name: "package_code_coverage",
+			Help: "Package code coverage.",
+		},
+		[]string{"package", "repo"},
+	)
+
+	repoRegister = map[string]*repo{}
+
+	codeCovTrackJob = "postsubmit"
```
Comment (sebastienvas) [inline toolbox/metrics/main.go]: I would make this configurable.

File: toolbox/metrics/main.go [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+const (
+	fetchInterval = 5
+)
+
+var (
+	webAddress = flag.String("listen_port", ":9103", "port on which to expose metrics and web interface.")
+	jenkinsURL = flag.String("jenkins_url", "https://testing.istio.io", "URL to Jenkins API")
+	gcsBucket  = flag.String("bucket", "istio-code-coverage", "gcs bucket")
+
+	succeededBuilds = prometheus.NewSummaryVec(
```
Comment (sebastienvas) [inline toolbox/metrics/main.go]: I feel this should go into a struct

File: toolbox/metrics/main.go [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+	)
+
+	repoRegister = map[string]*repo{}
+
+	codeCovTrackJob = "postsubmit"
+
+	gcsClient  *storage.Client
+	httpClient = &http.Client{}
+
+	jk = &jkConst{
```
Comment (sebastienvas) [inline toolbox/metrics/main.go:56]: Can you just declare those as const ?

Comment (yutongz) [reply toolbox/metrics/main.go:56]: Honestly, I prefer this way. In my point of view, these strings are from JSON files of Jenkins, so better put them into a structure.

**Comment by yutongz** [comment]
PTAL

File: toolbox/metrics/main.go [hunk 10]
```diff
@@ -0,0 +1,422 @@
...
+		name:               "name",
+		jobs:               "jobs",
+		building:           "building",
+		result:             "result",
+		duration:           "duration",
+		number:             "number",
+		resultFailure:      "FAILURE",
+		resultSUCCESS:      "SUCCESS",
```
Comment (sebastienvas) [inline toolbox/metrics/main.go]: rename to resultSuccess for consistency.

File: toolbox/metrics/main.go (hunk 9 above)
Comment (sebastienvas) [reply toolbox/metrics/main.go:56]: Golang does not support const struct which makes it ugly to declare a struct to put in const in.

**Review by sebastienvas — APPROVED** [review]

**Pull request merged**